src/utils/UVGenerator.js

```diff
@@ -0,0 +1,173 @@
+// https://github.com/mozilla/Spoke/commit/9701d647020e09d584885bd457eb225e9995c12f
+import XAtlas from '@gfodor/xatlas-web';
+import { Float32BufferAttribute, Uint32BufferAttribute } from 'three';
+
+const AddMeshStatus = {
+	Success: 0,
+	Error: 1,
+	IndexOutOfRange: 2,
+	InvalidIndexCount: 3
+};
+
+export class UVGenerator {
+
+	constructor() {
+
+		this._module = null;
+		this.channel = 1;
+
+	}
+
+	async init() {
+
+		if ( this._module ) return;
+
+		const wasmurl = new URL( '../../node_modules/@gfodor/xatlas-web/dist/xatlas-web.wasm', import.meta.url );
+		this._module = XAtlas( {
+
+			locateFile( path ) {
+
+				if ( path.endsWith( '.wasm' ) ) {
+
+					return wasmurl.toString();
+
+				}
+
+				return path;
+
+			}
+
+		} );
+
+		this._xatlas = await this._module;
+
+	}
+
+	generate( geometries, onProgress = null ) {
+
+		if ( geometries.constructor !== Array ) {
+
+			geometries = [ geometries ];
+
+		}
+
+		const xatlas = this._xatlas;
+		xatlas.createAtlas( );
+
+		const meshInfos = [];
+
+		for ( let i = 0; i < geometries.length; i ++ ) {
+
+			const geometry = geometries[ i ];
+
+			const originalVertexCount = geometry.attributes.position.count;
+			const originalIndexCount = geometry.index.count;
+
+			const meshInfo = xatlas.createMesh( originalVertexCount, originalIndexCount, true, true );
+			xatlas.HEAPU32.set( geometry.index.array, meshInfo.indexOffset / Uint32Array.BYTES_PER_ELEMENT );
+			xatlas.HEAPF32.set( geometry.attributes.position.array, meshInfo.positionOffset / Float32Array.BYTES_PER_ELEMENT );
+			xatlas.HEAPF32.set( geometry.attributes.normal.array, meshInfo.normalOffset / Float32Array.BYTES_PER_ELEMENT );
+
+			if ( geometry.attributes.uv ) {
+
+				xatlas.HEAPF32.set( geometry.attributes.uv.array, meshInfo.uvOffset / Float32Array.BYTES_PER_ELEMENT );
+
+			} else {
+
+				const vertCount = geometry.attributes.position.count;
+				geometry.setAttribute( 'uv', new Float32BufferAttribute( new Float32Array( vertCount * 2 ), 2, false ) );
+
+			}
+
+			const statusCode = xatlas.addMesh();
+			if ( statusCode !== AddMeshStatus.Success ) {
+
+				throw new Error( `UVUnwrapper: Error adding mesh. Status code ${ statusCode }` );
+
+			}
+
+			meshInfos.push( meshInfo );
+
+		}
+
+		const params = { padding: 2 };
+		if ( onProgress ) {
+
+			params.onProgress = onProgress;
+
+		}
+
+		xatlas.generateAtlas( params );
```

Comment:
I believe the original code that was in your repo for Spoke didn't use any of the options. I found a fork of the xatlas-web work done by Robert which added some of the parameters and upgraded xatlas, which I then made another derivative fork (now included as @gfodor/xatlas-web) which updated the emscripten build and exposed a few more parameters. The only parameter I actually needed here was padding, for the reasons mentioned above with edge expansion (with no padding, the edge expander would bleed color across islands), and so that was what prompted me to fork the original xatlas-web project.

Reply:
Looking again the `bruteForce` and `blockAlign` options might be the only other ones that could be worth exposing? We can do that another time, though.

> I believe the original code that was in your repo for Spoke didn't use any of the options.

Tbh I forgot about that 😅 It was added a long time ago but I got distracted by other work and never put much time into it since I didn't have an immediate use.